File: Table Importing/XRTableImporterDelimiterController.m
//
// MIT License
//
// Copyright (c) 2006 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
#import "XRTableImporterDelimiterController.h"

@implementation XRTableImporterDelimiterController
//
//+ (void)initialize {
//    [XRTableImporterDelimiterController setKeys:
//        [NSArray arrayWithObjects:@"delimiterPopup", nil]
//        triggerChangeNotificationsForDependentKey:@"activateDelimiterField"];
//}

+ (NSSet *)keyPathsForValuesAffectingValueForKey:(NSString *)key {
    
    NSSet *keyPaths = [super keyPathsForValuesAffectingValueForKey:key];
    
    if ([key isEqualToString:@"activateDelimiterField"]) {
        NSArray *affectingKeys = @[@"delimiterPopup"];
        keyPaths = [keyPaths setByAddingObjectsFromArray:affectingKeys];
    }
    return keyPaths;
}

-(id)initWithPath:(NSString *)aPath
{

	if (!(self = [super init])) return nil;

	if(self)
	{
        [[NSBundle mainBundle] loadNibNamed:@"XRTableInitialTextController"
                         owner:self
               topLevelObjects:nil];
		if(aPath)
			[self setValue:[[aPath lastPathComponent] stringByDeletingPathExtension] forKeyPath:@"tableName"];
		else
			[self setValue:@"Untitled_Table" forKeyPath:@"tableName"];
		[self setValue:@", " forKeyPath:@"delimiterFieldValue"];
        [self setValue:[NSNumber numberWithInt:NSControlStateValueOn] forKeyPath:@"columnTitles"];
		
	}
	
	return self;
}

- (IBAction)cancelImport:(id)sender
{
    [self.window.sheetParent endSheet:self.window returnCode:NSModalResponseCancel];
}

- (IBAction)import:(id)sender
{
    [self.window.sheetParent endSheet:self.window returnCode:NSModalResponseOK];
}

-(BOOL)activateDelimiterField
{
	if(delimiterPopup>0)
		return YES;
	else
		return NO;
}

-(NSWindow *)window
{
	return window;
}

-(NSDictionary *)results
{
	NSMutableDictionary *theDict = [[NSMutableDictionary alloc] init];

	[theDict setObject:[self valueForKeyPath:@"tableName"] forKey:@"tableName"];
	[theDict setObject:[self valueForKeyPath:@"columnTitles"] forKey:@"columnTitles"];
	[theDict setObject:[self valueForKeyPath:@"delimiterPopup"] forKey:@"delimiterPopup"];
	[theDict setObject:[self valueForKeyPath:@"delimiterFieldValue"] forKey:@"delimiterFieldValue"];
	
	switch([[self valueForKeyPath:@"encodingValue"] intValue])
	{
		case 0:
			[theDict setObject:[NSNumber numberWithInt:NSASCIIStringEncoding] forKey:@"encodingValue"];
			break;
		case 1:
			[theDict setObject:[NSNumber numberWithInt:NSMacOSRomanStringEncoding] forKey:@"encodingValue"];
			break;
		case 2:
			[theDict setObject:[NSNumber numberWithInt:NSWindowsCP1254StringEncoding] forKey:@"encodingValue"];
			break;
		case 3:
			[theDict setObject:[NSNumber numberWithInt:NSUTF8StringEncoding] forKey:@"encodingValue"];
			break;
		case 4:
			[theDict setObject:[NSNumber numberWithInt:NSISOLatin1StringEncoding] forKey:@"encodingValue"];
			break;
		case 5:
			[theDict setObject:[NSNumber numberWithInt:NSISOLatin2StringEncoding] forKey:@"encodingValue"];
			break;
		default:
			[theDict setObject:[NSNumber numberWithInt:NSASCIIStringEncoding] forKey:@"encodingValue"];
			break;
	}
	//NSLog([theDict description]);
	return [NSDictionary dictionaryWithDictionary:theDict];
}


@end

File: PaleoRose/Classes/Table Importing/Delimiter Controller/XRTableImporterDelimiterController.swift
// XRTableImporterDelimiterController.swift
//
// MIT License
//
// Copyright (c) 2006 to present Thomas L. Moore.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

import Cocoa

@objc class XRTableImporterDelimiterController: NSObject {
    // MARK: - Properties
    
    @objc private(set) var tableName: String = ""
    @objc private var columnTitles: Int = 0
    @objc private var delimiterFieldValue: String = ""
    @objc private var delimiterPopup: Int = 0
    @objc private var encodingValue: Int = 0
    
    @IBOutlet private weak var window: NSWindow!
    @IBOutlet private weak var tableNameField: NSTextField!
    @IBOutlet private weak var delimiterField: NSTextField!
    
    // MARK: - Initialization
    
    @objc init?(path: String) {
        super.init()
        
        guard Bundle.main.loadNibNamed("XRTableInitialTextController", owner: self, topLevelObjects: nil) else {
            return nil
        }
        
        if !path.isEmpty {
            tableName = (path as NSString).lastPathComponent.deletingPathExtension
        } else {
            tableName = "Untitled_Table"
        }
        
        delimiterFieldValue = ", "
        columnTitles = NSControl.StateValue.on.rawValue
    }
    
    // MARK: - Actions
    
    @IBAction func cancelImport(_ sender: Any) {
        window.sheetParent?.endSheet(window, returnCode: .cancel)
    }
    
    @IBAction func import(_ sender: Any) {
        window.sheetParent?.endSheet(window, returnCode: .OK)
    }
    
    // MARK: - Public API
    
    @objc var activateDelimiterField: Bool {
        delimiterPopup > 0
    }
    
    override class func keyPathsForValuesAffecting(_ key: String) -> Set<String> {
        var keyPaths = super.keyPathsForValuesAffecting(key)
        
        if key == "activateDelimiterField" {
            keyPaths.insert("delimiterPopup")
        }
        
        return keyPaths
    }
    
    @objc var results: [String: Any] {
        var dict: [String: Any] = [
            "tableName": tableName,
            "columnTitles": columnTitles,
            "delimiterPopup": delimiterPopup,
            "delimiterFieldValue": delimiterFieldValue
        ]
        
        let encoding: String.Encoding
        switch encodingValue {
        case 0:
            encoding = .ascii
        case 1:
            encoding = .macOSRoman
        case 2:
            encoding = .windowsCP1254
        case 3:
            encoding = .utf8
        case 4:
            encoding = .isoLatin1
        case 5:
            encoding = .isoLatin2
        default:
            encoding = .ascii
        }
        
        dict["encodingValue"] = encoding.rawValue
        return dict
    }
}
